mysql/resource_grant.go

```diff
@@ -688,20 +697,42 @@ func ImportGrant(ctx context.Context, d *schema.ResourceData, meta interface{})
 	return results, nil
 }
 
+func produceGrantId(database string, username string, host string) string {
+	if strings.Compare(database, "*") != 0 && !strings.HasSuffix(database, "`") {
+		reProcedure := regexp.MustCompile(`(?i)^(function|procedure) (.*)$`)
+		if reProcedure.MatchString(database) {
+			// This is only a hack - user can specify function / procedure as database.
+			database = reProcedure.ReplaceAllString(database, "$1 `${2}`")
+		} else {
+			database = fmt.Sprintf("`%s`", database)
+		}
+	}
+
+	return fmt.Sprintf("%s@%s:%s", database, username, host)
+}
+
 // setDataFromGrant copies the values from MySQLGrant to the schema.ResourceData
 // This function is used when importing a new Grant, or when syncing remote state to Terraform state
 // It is responsible for pulling any non-identifying properties (e.g. grant, tls_option) into the Terraform state
 // Identifying properties (database, table) are already set either as part of the import id or required properties
 // of the Terraform resource.
 func setDataFromGrant(grant MySQLGrant, d *schema.ResourceData) *schema.ResourceData {
+	userOrRole := grant.GetUserOrRole()
 	if tableGrant, ok := grant.(*TablePrivilegeGrant); ok {
 		d.Set("grant", grant.GrantOption())
 		d.Set("tls_option", tableGrant.TLSOption)
-
+		database := tableGrant.Database
+		id := produceGrantId(database, userOrRole.Name, userOrRole.Host)
+		d.SetId(id)
+		d.Set("database", database)
+		d.Set("table", tableGrant.Table)
```

Comment:
When we're at it, don't you need to set also user/host?

Reply:
This is already being set in [here](https://github.com/gallois/terraform-provider-mysql/blob/fix_grant_import/mysql/resource_grant.go#L760-L765). I feel like it makes sense to be done later, as it's common for all grant types and, being so, we can avoid some duplication

Reply:
Ah ok, I missed it.